04-onion-routing.md

```diff
@@ -875,6 +875,16 @@ The channel from the processing node has been disabled.
 
 The CLTV expiry in the HTLC is too far in the future.
 
+1. type: PERM|22 (`invalid_onion_payload`)
+2. data:
+   * [`varint`:`type`]
+   * [`u16`:`offset`]
+
+The decrypted onion per-hop payload was not understood by the processing node
+or is incomplete. If the failure can be narrowed down to a specific tlv type in
+the payload, the erring node may include that `type` and its byte `offset` in
```

Comment:
i missed the last spec meeting, so forgive me if i'm missing something, but atm this seems underspecified.

if we couldn't parse a record's type, what should `type` be? all possible values of `varint` correspond to valid types, so how can we distinguish this case?

also, in most cases i would think that `offset` is redundant information, as if the sender knows the invalid type they can inspect the payload they sent to figure out where the error is.

the main thing tho, is that it doesn't seem clear from this tho how the sender should choose between interpreting the index or the type.

Reply:
> If the failure can be narrowed down to a specific tlv type

So if you can't parse a record's type, just leave the default value here (0).
That's where the offset might be useful: you can indicate the starting bit where your parser failed to even extract a valid type.

I don't think sender should dynamically interpret this and take an action based on this: this error code indicates a bug in one of the nodes in the route. This is only meant to be logged for further investigation.

Does that make sense?

Reply:
Just my two cents since I wasn't in attendance.

There are five classes of failures I can think of:
 1. stream not canonical
 2. unknown required type
 3. unable to decode field (wrong encoding by sender or receiver)
 4. field was omitted or included improperly
 5. unable to parse type

IMO 1 should never happen, we have unit tests and test vectors to ensure this, plus it would be easy to decode your own payload and verify that it's not canonical. Either way, we only need the type to figure out where we went wrong.

For 2, 3, and 4 we already know the type, and that should be enough for the sender to narrow down where the error occurred.

If you squint tho, 5 is really just a variant of 3. The reason a type would fail to parse is likely due to the preceding record being read short/long. The only reason it wouldn't is if our varint encoding is messed up, which also shouldn't happen since we have comprehensive test vectors for BigSize as well.

Is it not sufficient to send the previously parsed type (or 0 if it occurs on the first type)? The sender can look for the next type in the stream >= to that value that is where the failure occurred. This works even if the type being parsed is 0 or 1.

I don't see much value being added by the index, given that we can identify the record in all of these cases just from the type and the payload that the sender included in the onion. If we are sorely missing it, we can always extend the message later and include it.

That being said, I think we could better specify as in the above which types to send back under which failure conditions. Is that something people are interested in? It'd be nice to have some sort of agreement as what the receiver should place in event of each failure, otherwise everyone needs to be aware of the nuisances of how other implementations pass back these debug values. Just adding them for the sake of logging isn't compelling as they're likely to get lost, however we could have actionable responses if we have some standardization, e.g. try to locate a newer node announcement or explain to the user that the node might not support a certain feature.

On a higher level, I am somewhat concerned in general about including the type at all, as it will allow people to fingerprint the set of required types a node supports assuming they themselves are properly encoding their payloads. An empty failure message would be preferable imo until the privacy implications have been considered more thoroughly.

Reply:
The consensus was honestly to not over-engineer this since the contents (`type` and `offset` are only optionally included in the error). This error shouldn't come up at all between valid implementations.
I agree that if your node supports required types that aren't in the spec that may allow probing that your node runs a specific version of a specific implementation. But feature bits likely let you do that probing as well so you're not learning anything new!